definitions/features/foreman_tasks.rb

```diff
@@ -22,8 +48,92 @@ def paused_tasks_count(ignored_tasks = [])
         WHERE state IN ('paused')
     SQL
     unless ignored_tasks.empty?
-      sql << "AND label NOT IN (#{ignored_tasks.map { |task| "'#{task}'" }.join(',')})"
+      sql << "AND label NOT IN (#{quotize(ignored_tasks)})"
     end
     feature(:foreman_database).query(sql).first['count'].to_i
   end
+
+  def count(query)
+    feature(:foreman_database).query(<<-SQL).first['count'].to_i
+     SELECT count(*) AS count FROM foreman_tasks_tasks WHERE #{condition(query)}
+    SQL
+  end
+
+  def delete(query)
+    tasks_condition = condition(query)
+
+    feature(:foreman_database).psql(<<-SQL)
+     BEGIN;
+       DELETE FROM dynflow_steps USING foreman_tasks_tasks WHERE (foreman_tasks_tasks.external_id = dynflow_steps.execution_plan_uuid) AND #{tasks_condition};
+       DELETE FROM dynflow_actions USING foreman_tasks_tasks WHERE (foreman_tasks_tasks.external_id = dynflow_actions.execution_plan_uuid) AND #{tasks_condition};
+       DELETE FROM dynflow_execution_plans USING foreman_tasks_tasks WHERE (foreman_tasks_tasks.external_id = dynflow_execution_plans.uuid) AND #{tasks_condition};
+       DELETE FROM foreman_tasks_tasks WHERE #{tasks_condition};
+     COMMIT;
+    SQL
+
+    count(query)
+  end
+
+  def condition(query)
+    raise 'Invalid Query' unless valid(query)
+
+    if query == :old
+      old_tasks_condition
+    else
+      tasks_condition(query)
+    end
+  end
+
+  private
+
+  def backup_dir(query)
+    "/var/lib/foreman/backup-tasks/#{query}/#{Time.now.strftime('%Y-%m-%d_%H-%M-%S')}"
+  end
+
+  def backup_table(table, query, fkey = 'execution_plan_uuid')
+    yield("Backup #{table} [running]")
+    sql = "SELECT #{table}.* FROM foreman_tasks_tasks JOIN #{table} ON\
+       (foreman_tasks_tasks.external_id = #{table}.#{fkey})"
+    export_csv(sql, "#{table}.csv", query)
+    yield("Backup #{table} [DONE]")
+  end
+
+  def export_csv(sql, file_name, query)
+    dir = prepare_for_backup(query)
+    execute("echo \"COPY (#{sql}) TO STDOUT WITH CSV;\" \
+      | su - postgres -c '/usr/bin/psql -d foreman' > #{dir}/#{file_name}")
+  end
+
+  def old_tasks_condition(state = "'stopped', 'paused'")
+    "foreman_tasks_tasks.state IN (#{state}) AND " \
+       "foreman_tasks_tasks.started_at < CURRENT_DATE - INTERVAL '#{MIN_AGE} days'"
+  end
+
+  def prepare_for_backup(query)
+    dir = backup_dir(query)
+    execute("mkdir -p #{dir}")
+    dir
+  end
+
+  def quotize(array)
+    array.map { |el| "'#{el}'" }.join(',')
+  end
+
+  def safe_to_delete_tasks_pending
+    PENDING
+  end
+
+  def safe_to_delete_tasks_planning
+    PLANNING
+  end
+
+  def tasks_condition(query)
+    safe_to_delete_tasks = quotize(send("safe_to_delete_tasks_#{query}"))
```

Comment:
The query feels like it should contain some kind of query, condition, but in facts it's more like state. We probably need a better naming for this. Also I thing we tasks that are safe to delete should not be distinquished by planning/pening.  It's more like a set of tasks that are safe to delete no matter in what state they are.

Another suggestion is avoiding the send and calling methods dynamically: it makes it pretty hard to read and navigate thought the code later: imagine one would like so find where is the `safe_to_delete_task_planning` method called from…